include/inexor/vulkan-renderer/wrapper/cpu_texture.hpp

```diff
@@ -0,0 +1,75 @@
+﻿#pragma once
+
+#include <string>
+#include <vector>
+
+namespace inexor::vulkan_renderer::wrapper {
+
+class CpuTexture {
+private:
+    const std::string m_file_name;
+    const std::string m_name;
+
+    int m_texture_width{0};
+    int m_texture_height{0};
+    int m_texture_channels{0};
+    int m_mip_levels{0};
+
+    std::vector<unsigned char> m_texture_data;
+
+    ///
+    void read_texture();
+
+    ///
+    void generate_error_texture_data();
+
+public:
+    /// @brief Creates a CpuTexture instance with a default texture.
+    CpuTexture();
+
+    /// @brief Reads a texture from a file.
+    /// @param file_name [in] The file name of the texture.
+    /// @param name [in] The internal memory allocation name of the texture.
+    CpuTexture(const std::string &file_name, const std::string &name);
+
+    CpuTexture(const CpuTexture &) = delete;
+    CpuTexture(CpuTexture &&) noexcept;
+    ~CpuTexture();
+
+    CpuTexture &operator=(const CpuTexture &) = delete;
+    CpuTexture &operator=(CpuTexture &&) = default;
+
+    [[nodiscard]] const std::string &file_name() const {
+        return m_file_name;
+    }
+
+    [[nodiscard]] const std::string &name() const {
+        return m_name;
+    }
+
+    [[nodiscard]] const int width() const {
+        return m_texture_width;
+    }
+
+    [[nodiscard]] const int height() const {
+        return m_texture_height;
+    }
+
+    [[nodiscard]] const int channels() const {
+        return m_texture_channels;
+    }
+
+    [[nodiscard]] const int mip_levels() const {
+        return m_mip_levels;
+    }
+
+    [[nodiscard]] void *data() const {
+        return const_cast<void *>(static_cast<const void *>(m_texture_data.data()));
```

Comment:
It is because `m_texture_data.data()` returns a const pointer because of the const in `void *data() const {`.  static_cast will not allow the const to be removed from the type.

I think I am going to add

`friend class GpuTexture;`

in the CpuTexture class declaration.  And then, we can remove this function entirely and it should hopefully be cleaner overall.

Reply:
Hmm, I would probably make the function non-const then, since the `const_cast` is breaking const-correctness. Not too sure about the friend thing, I feel like `CpuTexture` should probably be completely separate from `GpuTexture`, in case there are other users? But this definitely needs to be discussed further.

Reply:
On this section, I didn't know what to do.  I can't remove the const currently because of the following:

```
GpuTexture::GpuTexture(const wrapper::Device &device, const VkPhysicalDevice graphics_card,
                       const VmaAllocator vma_allocator, std::string name, const VkQueue data_transfer_queue,
                       const std::uint32_t data_transfer_queue_family_index, const CpuTexture &cpu_texture)
    : m_device(device), m_graphics_card(graphics_card), m_vma_allocator(vma_allocator), m_name(std::move(name)),
      m_data_transfer_queue(data_transfer_queue), m_data_transfer_queue_family_index(data_transfer_queue_family_index),
      m_copy_command_buffer(device, data_transfer_queue, data_transfer_queue_family_index) {

    m_texture_width = cpu_texture.width();
    m_texture_height = cpu_texture.height();
    m_texture_channels = cpu_texture.channels();
    m_mip_levels = cpu_texture.mip_levels();

    create_texture(cpu_texture.data(), cpu_texture.data_size());
}
```

Since CpuTexture is passed by const / ref, if the .data() call is not const, this will fail to compile.

So, I haven't made any changes around this yet; not sure what to do.

Reply:
If `CpuTexture` only stored the pointer that stb gives us (instead of a vector), I think this would solve the issue, since `CpuTexture` is not technically owning anything anymore.